File: src/queue.js

// http://www.amazon.com/Cracking-Coding-Interview-Programming-Questions/dp/098478280X
// Q 3.5
// http://courses.csail.mit.edu/iap/interview/Hacking_a_Google_Interview_Practice_Questions_Person_B.pdf
// Question: Queue Using Stacks 2014/4/13 16:49-16:52

var MyQueue = function() {
  this.stack_en = [];
  this.stack_de = [];
};

MyQueue.prototype.enqueue = function(val) {
  if (this.stack_de.length>0) {
    while(this.stack_de.length>0) {
      this.stack_en.push(this.stack_de.pop());
    }
  }
  this.stack_en.push(val);
};
MyQueue.prototype.dequeue = function() {
  while(this.stack_en.length>0) {
    this.stack_de.push(this.stack_en.pop());
  }
  return this.stack_de.pop();
};

var myQueue = new MyQueue();
myQueue.enqueue(1);
myQueue.enqueue(1);
myQueue.dequeue();
myQueue.enqueue(2);
myQueue.enqueue(2);
myQueue.dequeue();
myQueue.enqueue(3);
console.log(myQueue.dequeue(),myQueue.dequeue(),myQueue.dequeue(),myQueue.dequeue());

// http://www.amazon.com/Cracking-Coding-Interview-Programming-Questions/dp/098478280X
// Stack having min value

var MinStack = function() {
  this.vals = [];
  this.mins = [];
};
MinStack.prototype.push = function(val) {
  if (this.mins.length==0 || val<=this.mins[this.mins.length-1]) {
    this.mins.push(val);
  }
  this.vals.push(val);
};
MinStack.prototype.pop = function() {
  var val = this.vals.pop();
  if (val==this.mins[this.mins.length-1]) {
    this.mins.pop();
  }
  return val;
};
MinStack.prototype.min = function() {
  return this.mins[this.mins.length-1];
};

var minStack = new MinStack();
minStack.push(10);
minStack.push(5);
minStack.push(3);
minStack.push(3);
minStack.pop();
console.log(minStack.min());

// http://www.amazon.com/Cracking-Coding-Interview-Programming-Questions/dp/098478280X

var tower_of_hanoi = function() {
  var start = [4,3,2,1], buffer=[], target = [];
  function move(n, src, dest, buf) {
    if (n<1) return;
    move(n-1, src, buf, dest);
    dest.push(src.pop());
    console.log(start,buffer,target);
    move(n-1, buf, dest, src);
  }
  move(start.length, start, target, buffer);
};

tower_of_hanoi();

File: src/Queue.swift
//
//  Queue.swift
//  
//
//  Created by kori on 8/13/16.
//
//

import Foundation

/////////////////////////////////

/*
 
 https://www.amazon.com/Cracking-Coding-Interview-Programming-Questions/dp/098478280X
 Q 3.5
 
 enqueue 1, 2, 3, 4
 q: [1, 2, 3, 4]
 enS: [1, 2, 3, 4]
 deS: []
 
 dequeue
 q: [2, 3, 4]
 enS: []
 deS: [4, 3, 2, 1] -> [4, 3, 2]
 
 dequeue
 enS: []
 deS: [4, 3]
 
 enqueue 5
 enS: [3, 4, 5]
 deS: []
 
 */

class QueueByStacks {
  private var enqueStack: [Int] = []
  private var dequeStack: [Int] = []
  
  func enque(v: Int) {
    move(from: &dequeStack, to: &enqueStack)
    
    enqueStack.append(v)
  }
  
  func deque() -> Int? {
    move(from: &enqueStack, to: &dequeStack)
    
    guard let last = dequeStack.last else { return nil }
    dequeStack.removeLast()
    
    return last
  }
  
  private func move(inout from from: [Int], inout to: [Int]) {
    guard from.count > 0 else { return }
    
    while let last = from.last {
      from.removeLast()
      to.append(last)
    }
  }
}

let queueByStacks = QueueByStacks()
queueByStacks.enque(1)
queueByStacks.enque(2)
queueByStacks.enque(3)

assert(queueByStacks.deque()!==1)
